src/Microsoft.Restier.WebApi/Query/QueryTotalCount.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Restier.Core.Query;
+
+namespace Microsoft.Restier.WebApi.Query
+{
+    internal class QueryTotalCount : IQueryExecutor
+    {
+        public IQueryExecutor Inner { get; set; }
+
+        public async Task<QueryResult> ExecuteQueryAsync<TElement>(
+            QueryContext context,
+            IQueryable<TElement> query,
+            CancellationToken cancellationToken)
+        {
+            var countOption = context.ApiContext.GetApiService<ODataCountOption>();
+            if (countOption.IncludeTotalCount)
+            {
+                var countQuery = ExpressionHelpers.GetCountableQuery(query);
+                var expression = ExpressionHelpers.Count(countQuery.Expression, countQuery.ElementType);
```

Comment:
I still think reverse engineering the expression is not a perfect solution, it's prone to break if some changes made to WebApi query options builder. But it seems the only solution to save 1 more query.